### ISSUE TITLE
Import order is not invariant

### ISSUE THREAD

**Issue by dlangBugzillaToGithub** [dlang/dmd#19399]
### CC List

* Walter Bright (@WalterBright)
* Steven Schveighoffer (@schveiguy)

### Description

````markdown
Currently, compilation will change depending on the order that imports are processed.  This order is determined by the order that modules are passed to the compiler on the command line, and also the order they appear in each individual file.  Here are some examples:

Import order in file changes output:

--- foo.d
version (A)
{
    static import baz;
    static import bar;
}
version (B)
{
    static import bar;
    static import baz;
}
void main()
{
    import std.stdio;
    writeln(bar.thing);
    writeln(baz.thing);
}
--- bar.d
module baz;
enum thing = "this is baz from bar.d";
--- baz.d
module bar;
enum thing = "this is bar from baz.d";


dmd -version=A -run foo.d
----
this is bar from baz.d
this is bar from baz.d
----

dmd -version=B -run foo.d
----
this is baz from bar.d
this is baz from bar.d
----

Output changes depending on order of modules passed on the command line:

--- main.d
import ibar, ibaz;
void main()
{
    ibarfunc();
    ibazfunc();
}

--- ibar.d
import bar, std.stdio;
void ibarfunc()
{
    writeln(thing);
}

--- ibaz.d
import baz, std.stdio;
void ibazfunc()
{
    writeln(thing);
}

--- bar.d
module baz;
enum thing = "this is baz from bar.d";

--- baz.d
module bar;
enum thing = "this is bar from baz.d";


dmd ibar.d ibaz.d -run main.d
OUTPUT:
this is baz from bar.d
this is baz from bar.d

dmd ibaz.d ibar.d -run main.d
OUTPUT:
this is bar from baz.d
this is bar from baz.d

A PR to fix this has been submitted: https://github.com/dlang/dmd/pull/7900
````

**Comment by dlangBugzillaToGithub** [comment]
#### johnnymarler (@marler8997) commented on 2018-05-24T12:36:40Z



````markdown
Added 2 tests to dmd for this bug here: https://github.com/dlang/dmd/pull/8165 but the PR was rejected. Razvan doesn't think tests should be pushed to dmd by themselves (without their corresponding fixes).

> Razvan: I don't see much benefit in having tests that don't pass in the test suite. This puts a burden on the autotester and encourages people to make PRs just with tests creating a separation between the fix and the test.
````

**Comment by dlangBugzillaToGithub** [comment]
#### bugzilla (@WalterBright) commented on 2018-10-29T22:08:04Z



````markdown
Jonathan Marler's Relevant Pull Requests:

Fix module/import order invariance:

    https://github.com/dlang/dmd/pull/7900

Interpret imports the same regardless of other modules passed on the command line:

    https://github.com/dlang/dmd/pull/7878

Add broken import order invariance tests:

    https://github.com/dlang/dmd/pull/8165

Deprecate import/module name mismatch (ALL CASES):

    https://github.com/dlang/dmd/pull/7778
````

**Comment by dlangBugzillaToGithub** [comment]
#### schveiguy (@schveiguy) commented on 2018-10-30T00:21:16Z



````markdown
Hm... I didn't realize that the module identifier was not identical to the filename. That is interesting. It's not clearly stated in this bug report that this is a requirement to get the error to occur -- I thought it had something to do with the fact that `thing` appears in both files.

I would expect that if a module is loaded via using the filesystem, and the module name doesn't match how it was found, then this should be an error. Anything else seems to strain your sanity!

In other words, if you pass the modules on the command line, then they can have non-matching filenames, but if they were found via the module name mapping to directories, they had better match the module name.
````
